### PR TITLE
Ag cuda 12.1 more fixes

### PR DESCRIPTION
**Pull request by agibsonccc**
## What changes were proposed in this pull request?

Fixes misc test issues including:
1. diag scalar case (allows rank 0 arrays)
2. loss functions
3. reduction arrays don't throw exceptions on rank 0 arrays anymore
4. Delete tests with equal tostrings but unusual eps failures
## How was this patch tested?

Ran c++ tests

## Quick checklist

The following checklist helps ensure your PR is complete:

- [X ] Eclipse Contributor Agreement signed, and signed commits - see [IP Requirements](https://deeplearning4j.konduit.ai/multi-project/how-to-guides/contribute/eclipse-contributors) page for details
- [ X] Reviewed the [Contributing Guidelines](https://github.com/eclipse/deeplearning4j/blob/master/CONTRIBUTING.md) and followed the steps within.
- [ X] Created tests for any significant new code additions.
- [X ] Relevant tests for your changes are passing.

### CODE REVIEW

**Review by treo — CHANGES_REQUESTED** [review]
Once again, you've got quite a few stray prints all over the place.

**Comment by agibsonccc** [comment]
Did a quick commented code/print scan and removed quite a bit.

File: libnd4j/include/array/NDArray.hXX [hunk 1]
```diff
@@ -2489,9 +2494,11 @@ NDArray NDArray::asS() const {
   int numStrings = isScalar() ? 1 : lengthOf();
   auto data = bufferAsT<int8_t>() + offsetsLength;
   for (sd::LongType e = 0; e < numStrings; e++) {
+    sd_printf("e is currently %d\n",e);
     offsets[e] = dataLength;
     start = nInputoffsets[e];
     stop = nInputoffsets[e + 1];
+    sd_printf("Start is %d and stop is %d\n",start,stop);
```
Comment (treo) [inline libnd4j/include/array/NDArray.hXX]: still some more stray prints

File: libnd4j/include/array/NDArray.hXX [hunk 2]
```diff
@@ -2527,16 +2534,19 @@ NDArray NDArray::asS() const {
       auto cdata = outData + offsets[e];
       auto end = nInputoffsets[e + 1];
       auto idata = inData + nInputoffsets[e];
+      sd_printf("data type is %s e is %d offsets[%d] is %d and offsets[%d] is %d\n",DataTypeUtils::asString(dtype).c_str(),e,e,offsets[e],e+1,offsets[e+1]);
```
Comment (treo) [inline libnd4j/include/array/NDArray.hXX]: still some more stray prints

**Review by treo — CHANGES_REQUESTED** [review]

File: libnd4j/include/ops/declarable/helpers/cuda/svd.cu [hunk 3]
```diff
@@ -256,6 +201,7 @@ static void svdJcb(sd::LaunchContext* context, const NDArray* A, NDArray* S, NDA
   // U [m, m] or [m, n] if fullUV = false and m > n
   // V [n, n] or [n, m] if fullUV = false and m < n
 
+  sd_print("svdJcb\n");
```
Comment (treo) [inline libnd4j/include/ops/declarable/helpers/cuda/svd.cu]: yet another stray print.

File: libnd4j/include/ops/declarable/helpers/cuda/svd.cu [hunk 4]
```diff
@@ -264,171 +210,183 @@ static void svdJcb(sd::LaunchContext* context, const NDArray* A, NDArray* S, NDA
...
-    if (calcUV) {
-      pU = transA ? V : U;
-      pV = transA ? U : V;
+  int lda = transA ? pA->strideAt(0) : pA->strideAt(1);
+  int ldu(transA ? n : m), ldv(transA ? m : n);
+  bool uForder(true), vForder(true);
 
-      uForder = pU->sizeAt(0) == 1 || pU->strideAt(0) == 1;
-      vForder = pV->sizeAt(0) == 1 || pV->strideAt(0) == 1;
+  sd_print("Before calcUV\n");
```
Comment (treo) [inline libnd4j/include/ops/declarable/helpers/cuda/svd.cu]: and one more

File: libnd4j/include/ops/declarable/helpers/cuda/svd.cu [hunk 5]
```diff
@@ -264,171 +210,183 @@ static void svdJcb(sd::LaunchContext* context, const NDArray* A, NDArray* S, NDA
...
 
-      if (!uForder) {
-        pU = new NDArray(pU->dup('f'));
-        toDelete.push_back(pU);
-      }
+  if (calcUV) {
+    sd_print("before calcUV\n");
```
Comment (treo) [inline libnd4j/include/ops/declarable/helpers/cuda/svd.cu]: and more

File: libnd4j/include/ops/declarable/helpers/cuda/svd.cu [hunk 6]
```diff
@@ -264,171 +210,183 @@ static void svdJcb(sd::LaunchContext* context, const NDArray* A, NDArray* S, NDA
...
+  if (status != CUSOLVER_STATUS_SUCCESS) throw cuda_exception::build("svdJcb: cuda failed !", status);
 
-    if (status != CUSOLVER_STATUS_SUCCESS) throw cuda_exception::build("svdJcb: cuda failed !", status);
+  int* devInfo = nullptr;
+  const cusolverEigMode_t jobz = calcUV ? CUSOLVER_EIG_MODE_VECTOR : CUSOLVER_EIG_MODE_NOVECTOR;
+  const int econ = !fullUV;
+
+  if (transA) math::sd_swap<int>(m, n);
+
+  sd_print("before buffer\n");
```
Comment (treo) [inline libnd4j/include/ops/declarable/helpers/cuda/svd.cu]: and more

File: libnd4j/include/ops/declarable/helpers/cuda/svd.cu [hunk 7]
```diff
@@ -264,171 +210,183 @@ static void svdJcb(sd::LaunchContext* context, const NDArray* A, NDArray* S, NDA
...
+        *handle, jobz, econ, m, n, reinterpret_cast<float*>(pA->specialBuffer()), lda,
+        reinterpret_cast<float*>(pS->specialBuffer()),
+        calcUV ? reinterpret_cast<float*>(pU->specialBuffer()) : reinterpret_cast<float*>(nullPtr), ldu,
+        calcUV ? reinterpret_cast<float*>(pV->specialBuffer()) : reinterpret_cast<float*>(nullPtr), ldv, &lwork,
+        gesvdjParams);
+  else
+    THROW_EXCEPTION("svdJcb: given data type is unsupported !");
 
-    if (S->ews() != 1) S->assign(pS);
+  sd_print("Buffer size\n");
```
Comment (treo) [inline libnd4j/include/ops/declarable/helpers/cuda/svd.cu]: and more

File: libnd4j/include/ops/declarable/helpers/cuda/svd.cu [hunk 8]
```diff
@@ -264,171 +210,183 @@ static void svdJcb(sd::LaunchContext* context, const NDArray* A, NDArray* S, NDA
...
+        reinterpret_cast<float*>(dWork), lwork, devInfo, gesvdjParams);
+  } else
+    THROW_EXCEPTION("svdJcb: given data type is unsupported !");
+
+  sd_print("cuSolverDnSgesdvdj\n");
+  if (status != CUSOLVER_STATUS_SUCCESS) throw cuda_exception::build("svdJcb: cuda failed !", status);
+
+  manager.synchronize();
+
+  sd_print("manager synchronize\n");
```
Comment (treo) [inline libnd4j/include/ops/declarable/helpers/cuda/svd.cu]: and more

File: libnd4j/include/ops/declarable/helpers/cuda/svd.cu [hunk 9]
```diff
@@ -264,171 +210,183 @@ static void svdJcb(sd::LaunchContext* context, const NDArray* A, NDArray* S, NDA
...
+  if (calcUV) {
+    if (!uForder) U->assign(transA ? pV : pU);
+    if (!vForder) V->assign(transA ? pU : pV);
   }
+
+  if (!calcUV && m != n) delete arrToAvoidBugInAPI;
+
+  for (int i = toDelete.size() - 1; i >= 0; --i) delete toDelete[i];
+
+  sd_print("Before free\n");
```
Comment (treo) [inline libnd4j/include/ops/declarable/helpers/cuda/svd.cu]: and more

File: libnd4j/include/ops/declarable/helpers/cuda/svd.cu [hunk 10]
```diff
@@ -264,171 +210,183 @@ static void svdJcb(sd::LaunchContext* context, const NDArray* A, NDArray* S, NDA
...
+  if (devInfo) cudaFree(devInfo);
+  if (dWork) cudaFree(dWork);
+  if (gesvdjParams) cusolverDnDestroyGesvdjInfo(gesvdjParams);
+
+  sd_print("destroyed\n");
```
Comment (treo) [inline libnd4j/include/ops/declarable/helpers/cuda/svd.cu]: and more

**Review by treo — CHANGES_REQUESTED** [review]

**Comment by agibsonccc** [comment]
@treo  done

**Review by treo — APPROVED** [review]

**Pull request merged**